src/Language/PureScript/CoreFn/Convert/Datatypes.hs

```diff
@@ -0,0 +1,179 @@
+    mkDeclKind kargs = traverse (goKind . snd) kargs >>= \case
+      [] -> pure $ PIR.Type ()
+      xs -> pure $ foldr1 (PIR.KindArrow ()) (xs <> [PIR.Type ()])
+     where
+       goKind :: Maybe SourceType -> Either String (Kind ())
+       goKind mk = do
+         mkTy <- either (Left . prettyErrorT) pure $ traverse tryConvertType mk
+         mkKind mkTy
+
+    -- the arguments to the *type*, i.e., they're all tyvars
+    -- NOTE: We should really make changes such that `Maybe SourceType` is `SourceType`
+    mkArgDecl :: (Text, Maybe SourceType) -> Either String (TyVarDecl (Qualified (ProperName 'TypeName)) ())
+    mkArgDecl (varNm,mKind) = do
+      kind <- mkKind =<< either (Left . prettyErrorT) pure (traverse tryConvertType mKind)
+      -- Here, as elsewhere, a sourcePos qualified ident represents a bound var
+      let qVarNm = qualifyNull (ProperName varNm)
+      pure $ TyVarDecl () qVarNm kind
+
+type PIRTypeWithPSNames = PIR.Type (Qualified (ProperName 'TypeName)) PLC.DefaultUni ()
+
+toPIRType :: Ty -> Either String PIRTypeWithPSNames
+toPIRType = \case
+  IR.TyVar txt ->  pure $ PIR.TyVar () $ qualifyNull (ProperName txt)
+  TyCon qtn@(Qualified qb tn) -> case qb of
+    ByThisModuleName "Builtin" -> handleBuiltinTy qtn
+    ByThisModuleName "Prim" ->  handlePrimTy qtn
+    _ -> pure $ PIR.TyVar () qtn
+  -- TODO: Special handling for function types (-> is a TyCon in PS but a Ctor of the Type ADT in plc )
+  IR.TyApp t1 t2 -> goTypeApp t1 t2
+  Forall _ v mbk ty _ -> do
+    k <- mkKind mbk
+    ty' <- toPIRType ty
+    pure $ TyForall () (qualifyNull $ ProperName v) k ty'
+  other -> error $ "Upon reflection, other types like " <> ppTy other <> " shouldn't be allowed in the Ty ast"
+ where
+   goTypeApp (IR.TyApp f a) b
+     | f == TyCon C.Function = do
+         a' <- toPIRType a
+         b' <- toPIRType b
+         pure $ PIR.TyFun () a' b'
+     | otherwise = PIR.TyApp () <$> toPIRType a <*> toPIRType b
+
+   handleBuiltinTy :: Qualified (ProperName 'TypeName) -> Either String (PIR.Type tyname PLC.DefaultUni ())
+   handleBuiltinTy = \case
+     C.BuiltinData -> pure $ TyBuiltin () (PLC.SomeTypeIn PLC.DefaultUniData)
+     C.BuiltinPair -> pure $ TyBuiltin () (PLC.SomeTypeIn PLC.DefaultUniProtoPair)
+     C.BuiltinList -> pure $ TyBuiltin () (PLC.SomeTypeIn PLC.DefaultUniProtoList)
+     C.BuiltinByteString -> pure $ TyBuiltin () (PLC.SomeTypeIn PLC.DefaultUniByteString)
+     other -> Left $   "Error when translating to PIR types: unsupported builtin type: " <> show other
+
+   handlePrimTy = \case
+     C.Function -> Left $  "function types should be caught in apps"
+     C.Array    -> pure $ TyBuiltin () (PLC.SomeTypeIn PLC.DefaultUniProtoList) -- is this wrong? do we want a SOP list? too tired to know
+     C.String   -> pure $ TyBuiltin () (PLC.SomeTypeIn PLC.DefaultUniString)
+     C.Char     -> pure $ TyBuiltin () (PLC.SomeTypeIn PLC.DefaultUniInteger)
```

Comment:
Does UPLC have primitives for working with Char? unicode codepoint <-> symbol stuff? It might be a good idea to remove Char completely